runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,726 @@
+}
+
+/// Punishment hooks for disputes.
+pub trait PunishValidators {
+	/// Punish a series of validators who were for an invalid parablock. This is expected to be a major
+	/// punishment.
+	fn punish_for_invalid(session: SessionIndex, validators: impl IntoIterator<Item=ValidatorIndex>);
+
+	/// Punish a series of validators who were against a valid parablock. This is expected to be a minor
+	/// punishment.
+	fn punish_against_valid(session: SessionIndex, validators: impl IntoIterator<Item=ValidatorIndex>);
+
+	/// Punish a series of validators who were part of a dispute which never concluded. This is expected
+	/// to be a minor punishment.
+	fn punish_inconclusive(session: SessionIndex, validators: impl IntoIterator<Item=ValidatorIndex>);
+}
+
+pub trait Config:
+	frame_system::Config +
+	configuration::Config +
+	session_info::Config
+{
+	type Event: From<Event> + Into<<Self as frame_system::Config>::Event>;
+	type RewardValidators: RewardValidators;
+	type PunishValidators: PunishValidators;
+}
+
+decl_storage! {
+	trait Store for Module<T: Config> as ParaDisputes {
+		/// The last pruneed session, if any. All data stored by this module
+		/// references sessions.
+		LastPrunedSession: Option<SessionIndex>;
+		// All ongoing or concluded disputes for the last several sessions.
+		Disputes: double_map
+			hasher(twox_64_concat) SessionIndex,
+			hasher(blake2_128_concat) CandidateHash
+			=> Option<DisputeState<T::BlockNumber>>;
+		// All included blocks on the chain, as well as the block number in this chain that
+		// should be reverted back to if the candidate is disputed and determined to be invalid.
+		Included: double_map
+			hasher(twox_64_concat) SessionIndex,
+			hasher(blake2_128_concat) CandidateHash
+			=> Option<T::BlockNumber>;
+		// Maps session indices to a vector indicating the number of potentially-spam disputes
+		// each validator is participating in. Potentially-spam disputes are remote disputes which have
+		// fewer than `byzantine_threshold + 1` validators.
+		//
+		// The i'th entry of the vector corresponds to the i'th validator in the session.
+		SpamSlots: map hasher(twox_64_concat) SessionIndex => Option<Vec<u32>>;
+		// Whether the chain is frozen or not. Starts as `false`. When this is `true`,
+		// the chain will not accept any new parachain blocks for backing or inclusion.
+		// It can only be set back to `false` by governance intervention.
+		Frozen get(fn is_frozen): bool;
+	}
+}
+
+decl_event! {
+	pub enum Event {
+		/// A dispute has been initiated. The boolean is true if the dispute is local. \[para_id\]
+		DisputeInitiated(ParaId, CandidateHash, bool),
+		/// A dispute has concluded for or against a candidate. The boolean is true if the candidate
+		/// is deemed valid (for) and false if invalid (against).
+		DisputeConcluded(ParaId, CandidateHash, bool),
+		/// A dispute has timed out due to insufficient participation.
+		DisputeTimedOut(ParaId, CandidateHash),
```

Comment:
Governance mode, you can not say for sure what should be done, so do nothing and let gov take over.

Reply:
We'll slash everyone who should've participated in availability a tiny amount.  Interestingly, if we worried about disputes spam preventing real disputes, then we'll want correlated timeouts to increase up to 100%, but..  

Afaik, If we're confident the disputes spam blocking cannot obstruct real disputes because either honest parties votes in the spam dispute, which slashes someone adversarial, and slashes them lots due to the correlated slashing on bad disputes. 

We'd ideally spell this reasoning out more clearly I guess, but I felt Rob's spam design aligns nicely.

Reply:
If a dispute times out, it means that the nodes are unable to participate and recover availability. A dispute initiation is a claim that the data is available, so we apply a minor punishment to all nodes which participated in the dispute.